### ISSUE TITLE
Questions about your implementation

### ISSUE THREAD

**Issue by JJumSSu** [mrwu-mac/ControlMLLM#14]
Hi, first of all, thank you for your awesome work!

I am trying to use your implementation of attention visualization to visualize the attention map of a custom model.

I have the following questions about your method:

1. In README.md, you stated that `get_local` provides `gradient-based attention` and that gradients can be obtained by removing `torch.no_grad()`. However, I couldn't find any mention of `gradient-based attention` in your original paper. In the paper, you mentioned that you used an attention map rather than a relevancy map for implementation, as described in Section B.2. Could you clarify this? Also, I couldn't find using gradients for obtaining the attention map (i.e. relevancy map).

2.  I explored the "Visualizer" you implemented in the following [repository](https://github.com/mrwu-mac/Visualizer/tree/main). Since the README is written in Chinese, I translated it into English to better understand the details. Could you confirm if my understanding is correct?  

```Rather than manually registering hooks, which requires knowing the exact module path, your decorator-based visualization tool only needs to be applied to the attention block layer (e.g., `LlamaSdpaAttention`) of the target model. This would then extract all attention maps generated during the forward pass. Is that correct?```

Awesome work, thank you!

**Comment by mrwu-mac** [comment]
Hi, thank you for your interest in our work. Here’s a clarification of the points you raised:

1. Gradient-based Attention: In our implementation, gradients of attention maps are not directly used for visualization. Instead, gradients are leveraged during the latent optimization process to guide the learnable latent representations (this is unrelated to visualizing raw attention maps). If your goal is solely to visualize attention maps (without latent optimization), you can safely ignore gradients and use the raw attention weights generated during the forward pass.

2.Relevancy Maps: Since relevancy maps are not part of our current framework, the related code was not included in the open-source release.

3.Visualizer Tool: You are right. The decorator-based Visualizer in our code automatically extracts all attention maps from the decorated attention block layers (e.g., LlamaSdpaAttention) during the forward pass.

**Comment by JJumSSu** [comment]
Thank you for your kind response. It is very clear to me now.

Another question I have is that

I would like to experiment with relevance maps (attention maps with gradients) using your codebase with some modifications. Based on your explanation in Appendix B.2, my understanding is that to obtain gradients for generating a relevance map, I need to first generate the token, compute the loss, perform the backward operation, and then access the .grad variables associated with the attention mechanisms (e.g., LlamaSdpaAttention.forward). Is that correct?

**Comment by mrwu-mac** [comment]
Perhaps you can draw inspiration from this project (https://github.com/hila-chefer/Transformer-Explainability) and our previous paper "DIFNet: Boosting Visual Information Flow for Image Captioning" . To obtain relevance maps, the goal is to extract the relevance on each attention map for the output using the LRP algorithm after a single inference.

**Comment by JJumSSu** [comment]
Thank you for your response, I'll close the issue!
